### ISSUE TITLE
Mysql Join multiple Table error: not unique table/alias

### ISSUE THREAD

**Issue by 4dsistemi** [propelorm/Propel2#1783]
Hello,

I have two foreign keys in the same table connected to another table.
If I use the releatedby query I get the same error reported in the title.

the issue is use the joinwith query with the alias. There is another way to tho this?

Thanks

**Comment by mringler** [comment]
Hey @4dsistemi,
can you post the definitions of the two tables in your schema.xml, and the failing join statement?

**Comment by 4dsistemi** [comment]
Here you are. the problem is on son_id and customer_id

`<table name="isifit_spesa_pratica_sportiva_minori" idMethod="native" phpName="IsifitSpesaPraticaSportivaMinori" namespace="database">
    <column name="id" phpName="Id" type="INTEGER" primaryKey="true" autoIncrement="true" required="true"/>
    <column name="creation" phpName="Creation" type="TIMESTAMP" required="true"/>
    <column name="amount" phpName="Amount" type="FLOAT" required="true" defaultValue="0"/>
    <column name="isifit_sport_id" phpName="IsifitSportId" type="INTEGER" required="true"/>
    <column name="isifit_causale_pagamento_id" phpName="IsifitCausalePagamentoId" type="INTEGER" required="true"/>
    <column name="customer_id" phpName="CustomerId" type="INTEGER" required="true"/>
    <column name="son_id" phpName="SonId" type="INTEGER" required="true"/>
    <column name="commercial_local_id" phpName="CommercialLocalId" type="VARCHAR" size="50" required="true"/>
    <foreign-key foreignTable="commercial" name="fk_isifit_spesa_pratica_sportiva_minori_commercial1" onDelete="CASCADE" onUpdate="CASCADE">
      <reference local="commercial_local_id" foreign="local_id"/>
    </foreign-key>
    <foreign-key foreignTable="customer" name="fk_isifit_spesa_pratica_sportiva_minori_customer1">
      <reference local="customer_id" foreign="id"/>
    </foreign-key>
    <foreign-key foreignTable="customer" name="fk_isifit_spesa_pratica_sportiva_minori_customer2">
      <reference local="son_id" foreign="id"/>
    </foreign-key>
    <foreign-key foreignTable="isifit_causale_pagamento" name="fk_isifit_spesa_pratica_sportiva_minori_isifit_causale_pagame1">
      <reference local="isifit_causale_pagamento_id" foreign="id"/>
    </foreign-key>
    <foreign-key foreignTable="isifit_sport" name="fk_isifit_spesa_pratica_sportiva_minori_isifit_sport1">
      <reference local="isifit_sport_id" foreign="id"/>
    </foreign-key>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_commercial1_idx">
      <index-column name="commercial_local_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_customer1_idx">
      <index-column name="customer_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_customer2_idx">
      <index-column name="son_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_isifit_sport1_idx">
      <index-column name="isifit_sport_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_isifit_causale_paga_idx">
      <index-column name="isifit_causale_pagamento_id"/>
    </index>
    <vendor type="mysql">
      <parameter name="Engine" value="InnoDB"/>
    </vendor>
  </table>`

`<table name="customer" idMethod="native" phpName="Customer" namespace="database">
    <column name="id" phpName="Id" type="INTEGER" primaryKey="true" autoIncrement="true" required="true"/>
    <column name="name" phpName="Name" type="VARCHAR" size="45" required="true"/>
    <column name="surname" phpName="Surname" type="VARCHAR" size="45" required="true"/>
    <column name="iva" phpName="Iva" type="VARCHAR" size="50"/>
    <column name="email" phpName="Email" type="VARCHAR" size="45"/>
    <column name="address" phpName="Address" type="VARCHAR" size="45"/>
    <column name="city" phpName="City" type="VARCHAR" size="50" required="true"/>
    <column name="province" phpName="Province" type="VARCHAR" size="50" required="true"/>
    <column name="zip" phpName="Zip" type="VARCHAR" size="10" required="true"/>
    <column name="country" phpName="Country" type="VARCHAR" size="50" required="true"/>
    <column name="phone" phpName="Phone" type="VARCHAR" size="45"/>
    <column name="pec" phpName="Pec" type="VARCHAR" size="45"/>
    <column name="ae_code" phpName="AeCode" type="VARCHAR" size="45"/>
    <column name="birthday" phpName="Birthday" type="VARCHAR" size="45"/>
    <column name="society" phpName="Society" type="VARCHAR" size="45"/>
    <column name="fiscal" phpName="Fiscal" type="VARCHAR" size="45"/>
    <column name="birthplace" phpName="Birthplace" type="VARCHAR" size="50" required="true"/>
    <column name="gender" phpName="Gender" type="INTEGER" defaultValue="0"/>
    <column name="serial_list_local_id" phpName="SerialListLocalId" type="VARCHAR" size="50" required="true"/>
    <column name="active" phpName="Active" type="INTEGER" required="true" defaultValue="1"/>
    <column name="doc_type" phpName="DocType" type="VARCHAR" size="20"/>
    <column name="doc_serial" phpName="DocSerial" type="VARCHAR" size="50"/>
    <column name="doc_release" phpName="DocRelease" type="VARCHAR" size="40"/>
    <column name="doc_expire" phpName="DocExpire" type="DATE"/>
    <foreign-key foreignTable="commercial" name="fk_customer_serial_list1" onDelete="CASCADE" onUpdate="CASCADE">
      <reference local="serial_list_local_id" foreign="local_id"/>
    </foreign-key>
    <index name="fk_customer_serial_list1_idx">
      <index-column name="serial_list_local_id"/>
    </index>
    <vendor type="mysql">
      <parameter name="Engine" value="InnoDB"/>
    </vendor>
  </table>`

Here the query:

`IsifitSpesaPraticaSportivaMinoriQuery::create()
    ->innerJoinWithCustomerRelatedByCustomerId()
    ->innerJoinWithCustomerRelatedBySonId()
    ->innerJoinWithIsifitCausalePagamento()
    ->innerJoinWithIsifitSport()
    ->filterByCommercialLocalId(getCommercialFormSerial($serial))
    ->orderByCreation(Criteria::DESC)
    ->find();`

Thanks a lot

**Comment by 4dsistemi** [comment]
Here the tables:

<table name="customer" idMethod="native" phpName="Customer" namespace="database">
    <column name="id" phpName="Id" type="INTEGER" primaryKey="true" autoIncrement="true" required="true"/>
    <column name="name" phpName="Name" type="VARCHAR" size="45" required="true"/>
    <column name="surname" phpName="Surname" type="VARCHAR" size="45" required="true"/>
    <column name="iva" phpName="Iva" type="VARCHAR" size="50"/>
    <column name="email" phpName="Email" type="VARCHAR" size="45"/>
    <column name="address" phpName="Address" type="VARCHAR" size="45"/>
    <column name="city" phpName="City" type="VARCHAR" size="50" required="true"/>
    <column name="province" phpName="Province" type="VARCHAR" size="50" required="true"/>
    <column name="zip" phpName="Zip" type="VARCHAR" size="10" required="true"/>
    <column name="country" phpName="Country" type="VARCHAR" size="50" required="true"/>
    <column name="phone" phpName="Phone" type="VARCHAR" size="45"/>
    <column name="pec" phpName="Pec" type="VARCHAR" size="45"/>
    <column name="ae_code" phpName="AeCode" type="VARCHAR" size="45"/>
    <column name="birthday" phpName="Birthday" type="VARCHAR" size="45"/>
    <column name="society" phpName="Society" type="VARCHAR" size="45"/>
    <column name="fiscal" phpName="Fiscal" type="VARCHAR" size="45"/>
    <column name="birthplace" phpName="Birthplace" type="VARCHAR" size="50" required="true"/>
    <column name="gender" phpName="Gender" type="INTEGER" defaultValue="0"/>
    <column name="serial_list_local_id" phpName="SerialListLocalId" type="VARCHAR" size="50" required="true"/>
    <column name="active" phpName="Active" type="INTEGER" required="true" defaultValue="1"/>
    <column name="doc_type" phpName="DocType" type="VARCHAR" size="20"/>
    <column name="doc_serial" phpName="DocSerial" type="VARCHAR" size="50"/>
    <column name="doc_release" phpName="DocRelease" type="VARCHAR" size="40"/>
    <column name="doc_expire" phpName="DocExpire" type="DATE"/>
    <foreign-key foreignTable="commercial" name="fk_customer_serial_list1" onDelete="CASCADE" onUpdate="CASCADE">
      <reference local="serial_list_local_id" foreign="local_id"/>
    </foreign-key>
    <index name="fk_customer_serial_list1_idx">
      <index-column name="serial_list_local_id"/>
    </index>
    <vendor type="mysql">
      <parameter name="Engine" value="InnoDB"/>
    </vendor>
  </table>

<table name="isifit_spesa_pratica_sportiva_minori" idMethod="native" phpName="IsifitSpesaPraticaSportivaMinori" namespace="database">
    <column name="id" phpName="Id" type="INTEGER" primaryKey="true" autoIncrement="true" required="true"/>
    <column name="creation" phpName="Creation" type="TIMESTAMP" required="true"/>
    <column name="amount" phpName="Amount" type="FLOAT" required="true" defaultValue="0"/>
    <column name="isifit_sport_id" phpName="IsifitSportId" type="INTEGER" required="true"/>
    <column name="isifit_causale_pagamento_id" phpName="IsifitCausalePagamentoId" type="INTEGER" required="true"/>
    <column name="customer_id" phpName="CustomerId" type="INTEGER" required="true"/>
    <column name="son_id" phpName="SonId" type="INTEGER" required="true"/>
    <column name="commercial_local_id" phpName="CommercialLocalId" type="VARCHAR" size="50" required="true"/>
    <foreign-key foreignTable="commercial" name="fk_isifit_spesa_pratica_sportiva_minori_commercial1" onDelete="CASCADE" onUpdate="CASCADE">
      <reference local="commercial_local_id" foreign="local_id"/>
    </foreign-key>
    <foreign-key foreignTable="customer" name="fk_isifit_spesa_pratica_sportiva_minori_customer1">
      <reference local="customer_id" foreign="id"/>
    </foreign-key>
    <foreign-key foreignTable="customer" name="fk_isifit_spesa_pratica_sportiva_minori_customer2">
      <reference local="son_id" foreign="id"/>
    </foreign-key>
    <foreign-key foreignTable="isifit_causale_pagamento" name="fk_isifit_spesa_pratica_sportiva_minori_isifit_causale_pagame1">
      <reference local="isifit_causale_pagamento_id" foreign="id"/>
    </foreign-key>
    <foreign-key foreignTable="isifit_sport" name="fk_isifit_spesa_pratica_sportiva_minori_isifit_sport1">
      <reference local="isifit_sport_id" foreign="id"/>
    </foreign-key>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_commercial1_idx">
      <index-column name="commercial_local_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_customer1_idx">
      <index-column name="customer_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_customer2_idx">
      <index-column name="son_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_isifit_sport1_idx">
      <index-column name="isifit_sport_id"/>
    </index>
    <index name="fk_isifit_spesa_pratica_sportiva_minori_isifit_causale_paga_idx">
      <index-column name="isifit_causale_pagamento_id"/>
    </index>
    <vendor type="mysql">
      <parameter name="Engine" value="InnoDB"/>
    </vendor>
  </table>

And here the query:


IsifitSpesaPraticaSportivaMinoriQuery::create()
    ->innerJoinWithCustomerRelatedByCustomerId()
    ->innerJoinWithCustomerRelatedBySonId()
    ->innerJoinWithIsifitCausalePagamento()
    ->innerJoinWithIsifitSport()
    ->filterByCommercialLocalId(getCommercialFormSerial($serial))
    ->orderByCreation(Criteria::DESC)
    ->find();

Here my solution:

<foreign-key foreignTable="customer" name="fk_isifit_spesa_pratica_sportiva_minori_customer2" phpName="Son">
      <reference local="son_id" foreign="id"/>
    </foreign-key>

IsifitSpesaPraticaSportivaMinoriQuery::create()
    ->innerJoinWithCustomerRelatedByCustomerId()
    ->joinWith("Son s")
    ->innerJoinWithIsifitCausalePagamento()
    ->innerJoinWithIsifitSport()
    ->filterByCommercialLocalId(getCommercialFormSerial($serial))
    ->orderByCreation(Criteria::DESC)
    ->find();


thanks a lot

**Comment by mringler** [comment]
I think you need to set an alias for at least one of the customer tables in the join, i.e.:
```php
IsifitSpesaPraticaSportivaMinoriQuery::create()
->innerJoinWithCustomerRelatedByCustomerId('customer') // set table alias 'customer'
->innerJoinWithCustomerRelatedBySonId('son') // set table alias 'son'
->innerJoinWithIsifitCausalePagamento()
->innerJoinWithIsifitSport()
->filterByCommercialLocalId(getCommercialFormSerial($serial))
->orderByCreation(Criteria::DESC)
->find();
```

That's basically the same as you do in the second query:
```php
->joinWith("Son s") // use relation 'Son' and call the referenced customer table 's'
```
is the same as
```php
->joinWithCustomerRelatedBySonId('s')
```

Problem is that otherwise you end up with data from two tables both named "customer", and Propel can't figure out which is which. It could set an alias internally, but then you wouldn't know it and couldn't reference the columns. So the exception is the best Propel can do. It could be a bit more verbose though.

Long story short, try to add an alias as argument to at least one of your (inner) join statements. Does that work?
